OpenTimestamp client install & LND timestamp verification

VajraOfIndra opened the pull request:
#### What

This PR is a proposal to:
1) Install the OpenTimestamp client as a preparatory step before installing LND
2) Verify the LND binary timestamp with the above-mentionned client
3) Re-organise both the Bitcoin Core and LND install sections by following the following logical sequence:

~~**Download -> Signature check -> timestamp check -> checksum check -> Installation**~~
(New, following reviews): **Download -> Checksum check -> Signature check -> Timestamp check -> Installation**

- **Download** -> to download the required binary, checksum, signature and timestamp files
- **Checksum check** -> to verify the checksum of the binary
- **Signature check** -> to verify the signature of the checksum file just used above
- **Timestamp check** -> to verify the timestamp of the checksum file just used above
- **Installation** -> to install the program once the integrity of the downloaded binary file has been verified

_Note: if the PR is too confusing I could remove the modifictions done on `bitcoin-client.md` and focus this PR on LND only, let me know._

### Why

Verifying the timestamps adds an extra layer of security.

#### Scope

- [x] significant change to core configuration
- [ ] independent bonus guide
- [ ] simple bug fix

Fixes #1043 

#### Test & maintenance

* Follow new instructions to install client and verify the timestamp of LND
* Check that the instructions are flowing nicely and are not too long while still explaining what's going on.

![stamp](https://media.giphy.com/media/xT9DPHsT7jrLmLvw08/giphy-downsized.gif)

kristapsk commented:
Is there reason to do signature check before checksum check? Recently I changed JoinMarket installer to do the opposite, otherwise user got confused with signature validation error, when reason was incomplete download. https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1339

VajraOfIndra commented:
> Is there reason to do signature check before checksum check?

Just what I thought was more logical, i.e. before using the checksum file to verify the binary it seems ogial to verify first the authenticity of the checksum file.

But if it leads to potential confusions with edge cases (e.g. incomplete downloads), then we can do the checksum check first if you prefer.

kristapsk commented:
> Not sure what you mean.., As it is now, it's not for "lnd" user but the "admin" user. Would adding sudo make it available to other users than admin?

Yes, right. LND and bitcoind is installed from admin user not their respective users. But no so with JoinMarket and probably others in future.

When you do `pip3 install opentimestamps-client`, it will install in `/home/admin/.local/bin/`, when you do `sudo pip3 install opentimestamps-client`, it will install in `/usr/local/bin/` and will be available to all users.

Also, I think that OTS client installation should be moved to Bitcoin guide. For now it is added mainly because of LND, but this will change in future.

VajraOfIndra commented:
> When you do `pip3 install opentimestamps-client`, it will install in `/home/admin/.local/bin/`, when you do `sudo pip3 install opentimestamps-client`, it will install in `/usr/local/bin/` and will be available to all users.

Thanks for the clarification.

> Also, I think that OTS client installation should be moved to Bitcoin guide. For now it is added mainly because of LND, but this will change in future.

Where exactly in the Bitcoin guide would you like to have the instructions on how to install the client? and why would it make more sense since we can't use it at first install? 

For the Bitcoin Core updates, we could simply add a note and the command to use in the Update section at the end.

kristapsk commented:
> Where exactly in the Bitcoin guide would you like to have the instructions on how to install the client? and why would it make more sense since we can't use it at first install?

Could be between "Running bitcoind" and "Bitcoin Core is syncing" maybe. My idea is to have it installed at the beginning, directly after bitcoind, so that we have it installed system-wide and it is available for any following guide. Not only LND, but for anything else that may start timestamping releases (or actually RaspiBolt users might use it to timestamp any documents they may want for themselves too). No point in doing separate guide IMHO, as it  basically is just some description what it is and single command `sudo pip3 install opentimesamps-client`. If we add this to LND guide, some people may choose to use CLN instead and later, in JoinMarket guide, they will have OTS client missing. I don't say that my idea is the best approach, I'm open to other ideas, just this today seems to me the most logical one.

VajraOfIndra commented:
> If we add this to LND guide, some people may choose to use CLN instead and later

Excellent point about CLN, I didn't think about that.

Moving it at the end of the Bitcoin Core chapter sounds good indeed. The other solution that would also makes sense is to move it at the start when we install a bunch of program that we will need later:
![image](https://user-images.githubusercontent.com/86242283/190104147-ebf1d819-0982-4a6b-8a3b-e1504ff08c64.png)

We could add a line after that one to install the OTS client. However, I wonder if installing it before Bitcoin Core is running might mess up the configuration of the OTS client?

Anyways, will move it to the Bitcoin Core section now. PTAL.

VajraOfIndra commented:
> [...]

Ok, I updated the following:

1. I moved the OTS client installation to the Bitcoin Core guide, at the end, just before the user moves to the next section
2. I shifted the order of the checks, now starting with the checksum check, followed by the signature and ots check
3. I added a note in the "Update" section of Bitcoin Core to explain how to use the client instead of the website
4. I tired to explain a bit what's going on and the rationales of why we do this

PTAL and let me know if things can be simplified or clarified.
Feel free to reword entire sentences or sections, I don't mind. Clear wording is not my best skill.

kristapsk commented on guide/lightning/lightning-client.md:
```diff
@@ -24,19 +24,33 @@ We set up [LND](https://github.com/lightningnetwork/lnd/blob/master/README.md){:
   $ wget https://github.com/lightningnetwork/lnd/releases/download/v0.15.0-beta/lnd-linux-arm64-v0.15.0-beta.tar.gz
   $ wget https://github.com/lightningnetwork/lnd/releases/download/v0.15.0-beta/manifest-v0.15.0-beta.txt
   $ wget https://github.com/lightningnetwork/lnd/releases/download/v0.15.0-beta/manifest-roasbeef-v0.15.0-beta.sig
+  $ wget https://github.com/lightningnetwork/lnd/releases/download/v0.15.1-beta/manifest-roasbeef-v0.15.1-beta.sig.ots
```
```suggestion
  $ wget https://github.com/lightningnetwork/lnd/releases/download/v0.15.0-beta/manifest-roasbeef-v0.15.0-beta.sig.ots
```

kristapsk commented on guide/lightning/lightning-client.md:
````diff
@@ -59,12 +73,23 @@ We'll download, verify and install LND.
   ```sh
-  $ sha256sum --check manifest-v0.15.0-beta.txt --ignore-missing
-  > lnd-linux-arm64-v0.15.0-beta.tar.gz: OK
+  $ ots verify manifest-roasbeef-v0.15.1-beta.sig.ots
+  > [...]
+  > Success! Bitcoin block 751933 attests existence as of 2022-08-31 BST
````
```suggestion
  $ ots verify manifest-roasbeef-v0.15.0-beta.sig.ots
  > [...]
  > Success! Bitcoin block 742091 attests existence as of 2022-06-24 EEST
```

twofaktor commented on guide/bitcoin/bitcoin-client.md:
````diff
@@ -470,6 +470,22 @@ We also now want to enable the node to listen to and relay transactions.
+  ```sh
+  $ sudo pip3 install opentimestamps-client
+  ```
+
````
What do you think about adding lines 501 to 513 here?
It would be useful to check if Openstampclient is running properly and like a second and secure finally verification of Bitcoin Core binaries for yourself. Wdyt?

VajraOfIndra replied:
IMHO this is not really needed as we already verified the binaries above and would make a long guide even longer.

Ots being installed with apt should work fine.  But if you want we could add a couple of lines suchas 

> * Check that the OpenTimestamps client is properly installed
> 
>   ```sh
>   $ ots --version
>   ```

or even

> * Check that the OpenTimestamps client is properly installed

>   ```sh
>   $ ots --help
>   ```


Wdyt?

VajraOfIndra replied:
I added the version check.

Python 3 is the default version in RaspBerry Pi OS 64-bit Lite, so I think `pip3` is also default.

> `sudo apt-get install python3 python3-dev python3-pip python3-setuptools python3-wheel`

Are you suggetsing we add this line to the guide?

> is ready to merge?

Technically yes, it's been tested and approved. 

But I feel we lack some sort of consensus to add these timestamp checks... so I hesitate to merge it right now. I'd prefer if more maintainers and other contributors or users would express their support for it (so far @kristapsk who proposed this enhancment in #1043 , myself and you).

I wonder if it's up to us to better explain what attack this extra step protects against... a bit more education around what each step of the process really help with, i.e. 1) verifying the checksum of the binary for integrity (protects against mitm attacks), 2) verifying the signature of the checksums file for authenticity/web of trust (protects against hacked repos with malicious binaries and checksums), and finally 3) verifying the timestamp also for authenticity (protects against exposed private keys of signers, e.g. hacked PGP private key that happened after the release).

I think this would do a great FAQ entry. SImulations of real life hacks that could affect the Bitcoin Core binaries and how each step of the verification process expose each attacks. Anyways, this is for another discussion! :)

Point is, there seem to be little interest for this feature so far.. Wdyat?

VajraOfIndra replied:
Worked directly for me... but could be because I had already installed them before for another program..

VajraOfIndra replied:
The [PyPi OTS webpage](https://pypi.org/project/opentimestamps-client/) states that the only requirement to install OTS via PyPi is to have Python 3.

The [PyPi documentation](https://packaging.python.org/en/latest/tutorials/installing-packages/) states that to install a package there are only 2 requirements: Python 3 and pip are installed.

However they also advise to upgrade pip, setuptools and wheel with: `python3 -m pip install --upgrade pip setuptools wheel` but it's not a _necessary_ step.

Python 3 comes as part of the [packages in RasPiOS 64-bit Lite](https://downloads.raspberrypi.org/raspios_lite_arm64/images/raspios_lite_arm64-2022-09-26/2022-09-22-raspios-bullseye-arm64-lite.info). 

python3-pip is not listed in it though so maybe we need to install it?

The [PyPI documentation](https://packaging.python.org/en/latest/guides/installing-using-linux-tools/#debian-ubuntu) advises the following commands to install pip/setuptools/wheel:

> sudo apt update
> sudo apt install python3-venv python3-pip

So I propose to add these installation steps as preparatory steps to the guide just to be sure that pip is available and also to make sure pip, setuptools and wheel are up-tp-date:

> * Install the necessary Python 3 packages and make sure they are up-to-date
> > $ sudo apt update
> > $ sudo apt install python3-venv python3-pip
> > $ python3 -m pip install --upgrade pip setuptools wheel
> 
> * With user "admin", globally install the OpenTimestamp client
> > $ sudo pip3 install opentimestamps-client
> 
> * Display the OpenTimestamps client version to check that it is properly installed
> > $ ots --version

Wdyt? would that be a correct and robust way to install this Python package?

twofaktor reviewed (commented):
Feedback

twofaktor commented on guide/lightning/lightning-client.md:
````diff
@@ -59,12 +73,23 @@ We'll download, verify and install LND.
   ```sh
-  $ sha256sum --check manifest-v0.15.0-beta.txt --ignore-missing
-  > lnd-linux-arm64-v0.15.0-beta.tar.gz: OK
+  $ ots verify manifest-roasbeef-v0.15.0-beta.sig.ots
````
Change to:

`ots verify manifest-roasbeef-v0.15.0-beta.sig.ots -f manifest-roasbeef-v0.15.0-beta.sig`

official reference repository:

https://github.com/lightningnetwork/lnd/releases/tag/v0.15.0-beta

`Verifying the Release Timestamp` chapter

VajraOfIndra replied:
Good idea, changed.

twofaktor reviewed (changes requested):
Few suggestions

VajraOfIndra commented:
FYI; more information on how OpenTimestamps work (from 2016 but still informative): [https://petertodd.org/2016/opentimestamps-announcement#fn:hello-world-address](https://petertodd.org/2016/opentimestamps-announcement#fn:hello-world-address)

VajraOfIndra commented:
On top of the on-going discussion about package requirements to install ots (see [above](https://github.com/raspibolt/raspibolt/pull/1087#discussion_r990833930)), I have 2 questions:

**1. Virtual environment or pipx**

It is often recommended to install a Python program in a virtual environment rather than globally to avoid conflicts.
However, in their documentation, they rather recommend [pipx](https://packaging.python.org/en/latest/guides/installing-stand-alone-command-line-tools/) for standalone command line tools.. which seem to fit ots.

> Python “Virtual Environments” allow Python packages to be installed in an isolated location for a particular application, rather than being installed globally. If you are looking to safely install global command line tools, see Installing stand alone command line tool. [(source)](https://packaging.python.org/en/latest/tutorials/installing-packages/#installing-from-pypi)

Wdyat? should we use `pipx` to install the `opentimestamp-client` package? or a virtual env?

**2. How does OTS talk to our Bitcoin node..??!**

How come we don't have to provide information in a conf file about where our Bitcoin Core node runs on the host? How does OTS intreact with our Bitcoin node..? can't find much information so far, it's puzzling me.

sig-noi reviewed (changes requested):
`$ sudo pip3 install opentimestamps-client` works for me. OTS is now mentioned enough in the wild that I think it's worth including in the guide proper. I have only run RaspiBolt commands on this machine AFAIK so dependencies shouldn't be an issue. 

@VajraOfIndra, I appreciate wanting to not make the guide more complex if not necessary. However, to the extent that RaspiBolt doesn't explain OTS by including it I think we risk appearing incomplete and also confusing folks who wonder what it is/ why it's missing if they come across OTS elsewhere. So I vote to merge.

Prior to merge, the lightning-client.md section should be removed or updated to work with v0.15.3-beta. Not sure how to best integrate this with existing LND v0.15.3-beta pull request.

VajraOfIndra commented:
> `$ sudo pip3 install opentimestamps-client` works for me.

👍 

> @VajraOfIndra, I appreciate wanting to not make the guide more complex if not necessary. 

I'm personally for including it, but didn't want to merge my PR until there were a bit more support from the community and the other maintainers.

> Prior to merge, the lightning-client.md section should be removed or updated to work with v0.15.3-beta. Not sure how to best integrate this with existing LND v0.15.3-beta pull request.

Good point, I'll wait for the LND v0.15.3 PR to be merged and will update this PR accordingly.

@sig-noi Wdyt about the OTS client running as is without any confoguration.. does it seem normal to you? I feel we might be missing a few install/configuration steps. Heck, how does the OTS client knows where our Bitcoin Core client is? and how can it talk to it without special permissions? :)

kristapsk commented:
> Heck, how does the OTS client knows where our Bitcoin Core client is?

It looks at `~/.bitcoin/bitcoin.conf` - https://github.com/petertodd/python-bitcoinlib/blob/c1eca007724dc12fa1abfbc6df06965c703ad748/bitcoin/rpc.py#L148.

> and how can it talk to it without special permissions?

User using OTS should be in `bitcoin` group so that `/data/bitcoin/.cookie` is accessible.

VajraOfIndra commented:
> It looks at `~/.bitcoin/bitcoin.conf` - https://github.com/petertodd/python-bitcoinlib/blob/c1eca007724dc12fa1abfbc6df06965c703ad748/bitcoin/rpc.py#L148.

> User using OTS should be in `bitcoin` group so that `/data/bitcoin/.cookie` is accessible.

Thank you for the clarifications. So that's what it works out of the box with our setup 👍

twofaktor commented:
> > It looks at `~/.bitcoin/bitcoin.conf` - https://github.com/petertodd/python-bitcoinlib/blob/c1eca007724dc12fa1abfbc6df06965c703ad748/bitcoin/rpc.py#L148.
> 
> > User using OTS should be in `bitcoin` group so that `/data/bitcoin/.cookie` is accessible.

I opened an issue/help about this I opened an issue/help about this: https://github.com/opentimestamps/opentimestamps-client/issues/130

I'm waiting for a response: Wdyt?

> Thank you for the clarifications. So that's what it works out of the box with our setup 👍

kristapsk commented:
> > > It looks at `~/.bitcoin/bitcoin.conf` - https://github.com/petertodd/python-bitcoinlib/blob/c1eca007724dc12fa1abfbc6df06965c703ad748/bitcoin/rpc.py#L148.
> > 
> > 
> > > User using OTS should be in `bitcoin` group so that `/data/bitcoin/.cookie` is accessible.
> 
> I opened an issue/help about this: [opentimestamps/opentimestamps-client#130](https://github.com/opentimestamps/opentimestamps-client/issues/130)
> 
> I'm waiting for a response

No need for issue, it is working, I even pointed out to the exact lines of code that supports it.

twofaktor commented:
> > > > It looks at `~/.bitcoin/bitcoin.conf` - https://github.com/petertodd/python-bitcoinlib/blob/c1eca007724dc12fa1abfbc6df06965c703ad748/bitcoin/rpc.py#L148.
> > > 
> > > 
> > > > User using OTS should be in `bitcoin` group so that `/data/bitcoin/.cookie` is accessible.
> > 
> > 
> > I opened an issue/help about this: [opentimestamps/opentimestamps-client#130](https://github.com/opentimestamps/opentimestamps-client/issues/130)
> > I'm waiting for a response
> 
> No need for issue, it is working, I even pointed out to the exact lines of code that supports it.

We must avoid queries to public calendars if we already have our own node. Therefore, avoid the following logs:

  ```sh
  $ ots verify manifest-roasbeef-v0.15.3-beta.sig.ots -f manifest-roasbeef-v0.15.3-beta.sig
  > Got 1 attestation(s) from https://bob.btc.calendar.opentimestamps.org
  > Got 1 attestation(s) from https://finney.calendar.eternitywall.com
  > Got 1 attestation(s) from https://alice.btc.calendar.opentimestamps.org
  > Got 1 attestation(s) from https://btc.calendar.catallaxy.com
  > Success! Bitcoin block 759141 attests existence as of 2022-10-17 UTC
  ```

I misunderstand you?

twofaktor commented:
> > We must avoid queries to public calendars if we already have our own node.
> 
> You cannot avoid querying calendar servers, Bitcoin node does not have OTS merkle trees. Best you can maybe do here - use Tor for that. These are two different things - for `ots verify` you need **both** to query calendar servers and have full node.

ok, how can we use tor for this?

kristapsk commented:
> ok, how can we use tor for this? e.g: ` ots verify manifest-roasbeef-v0.15.3-beta.sig.ots -f manifest-roasbeef-v0.15.3-beta.sig --socks5-proxy 127.0.0.1:9050`
> 
> ??

Yes, that's the idea probably, but doesn't work for me. :( Problem could be that it tries to use SOCKS5 proxy both for connecting to calendar servers and Bitcoin node.
```
 ₿ ots --socks5-proxy 127.0.0.1:9050 verify manifest-roasbeef-v0.15.0-beta.sig.ots -f manifest-roasbeef-v0.15.0-beta.sig
Got 3 attestation(s) from cache
Traceback (most recent call last):
  File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 809, in connect
    negotiate(self, dest_addr, dest_port)
  File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 443, in _negotiate_SOCKS5
    self.proxy_peername, self.proxy_sockname = self._SOCKS5_request(
  File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 533, in _SOCKS5_request
    raise SOCKS5Error("{:#04x}: {}".format(status, error))
socks.SOCKS5Error: 0x01: General SOCKS server failure

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/admin/.local/bin/ots", line 8, in <module>
    sys.exit(main())
  File "/home/admin/.local/lib/python3.8/site-packages/otsclient/ots.py", line 34, in main
    args.cmd_func(args)
  File "/home/admin/.local/lib/python3.8/site-packages/otsclient/cmds.py", line 490, in verify_command
    if not verify_timestamp(detached_timestamp.timestamp, args):
  File "/home/admin/.local/lib/python3.8/site-packages/otsclient/cmds.py", line 412, in verify_timestamp
    block_count = proxy.getblockcount()
  File "/home/admin/.local/lib/python3.8/site-packages/bitcoin/rpc.py", line 489, in getblockcount
    return self._call('getblockcount')
  File "/home/admin/.local/lib/python3.8/site-packages/bitcoin/rpc.py", line 234, in _call
    self.__conn.request('POST', self.__url.path, postdata, headers)
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1011, in _send_output
    self.send(msg)
  File "/usr/lib/python3.8/http/client.py", line 951, in send
    self.connect()
  File "/usr/lib/python3.8/http/client.py", line 922, in connect
    self.sock = self._create_connection(
  File "/home/admin/.local/lib/python3.8/site-packages/otsclient/args.py", line 127, in create_connection
    sock.connect(address)
  File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 47, in wrapper
    return function(*args, **kwargs)
  File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 814, in connect
    raise GeneralProxyError("Socket error", error)
socks.GeneralProxyError: Socket error: 0x01: General SOCKS server failure
```

P.S. We could use Tor SOCKS5 proxy also for all downloads from GitHub.

twofaktor commented:
Same for me...

Can you think of what could be the solution? open an issue to https://github.com/opentimestamps/opentimestamps-client maybe?

> > ok, how can we use tor for this? e.g: ` ots verify manifest-roasbeef-v0.15.3-beta.sig.ots -f manifest-roasbeef-v0.15.3-beta.sig --socks5-proxy 127.0.0.1:9050`
> > ??
> 
> Yes, that's the idea probably, but doesn't work for me. :( Problem could be that it tries to use SOCKS5 proxy both for connecting to calendar servers and Bitcoin node.
> 
> ```
>  ₿ ots --socks5-proxy 127.0.0.1:9050 verify manifest-roasbeef-v0.15.0-beta.sig.ots -f manifest-roasbeef-v0.15.0-beta.sig
> Got 3 attestation(s) from cache
> Traceback (most recent call last):
>   File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 809, in connect
>     negotiate(self, dest_addr, dest_port)
>   File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 443, in _negotiate_SOCKS5
>     self.proxy_peername, self.proxy_sockname = self._SOCKS5_request(
>   File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 533, in _SOCKS5_request
>     raise SOCKS5Error("{:#04x}: {}".format(status, error))
> socks.SOCKS5Error: 0x01: General SOCKS server failure
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/admin/.local/bin/ots", line 8, in <module>
>     sys.exit(main())
>   File "/home/admin/.local/lib/python3.8/site-packages/otsclient/ots.py", line 34, in main
>     args.cmd_func(args)
>   File "/home/admin/.local/lib/python3.8/site-packages/otsclient/cmds.py", line 490, in verify_command
>     if not verify_timestamp(detached_timestamp.timestamp, args):
>   File "/home/admin/.local/lib/python3.8/site-packages/otsclient/cmds.py", line 412, in verify_timestamp
>     block_count = proxy.getblockcount()
>   File "/home/admin/.local/lib/python3.8/site-packages/bitcoin/rpc.py", line 489, in getblockcount
>     return self._call('getblockcount')
>   File "/home/admin/.local/lib/python3.8/site-packages/bitcoin/rpc.py", line 234, in _call
>     self.__conn.request('POST', self.__url.path, postdata, headers)
>   File "/usr/lib/python3.8/http/client.py", line 1256, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/usr/lib/python3.8/http/client.py", line 1011, in _send_output
>     self.send(msg)
>   File "/usr/lib/python3.8/http/client.py", line 951, in send
>     self.connect()
>   File "/usr/lib/python3.8/http/client.py", line 922, in connect
>     self.sock = self._create_connection(
>   File "/home/admin/.local/lib/python3.8/site-packages/otsclient/args.py", line 127, in create_connection
>     sock.connect(address)
>   File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 47, in wrapper
>     return function(*args, **kwargs)
>   File "/home/admin/.local/lib/python3.8/site-packages/socks.py", line 814, in connect
>     raise GeneralProxyError("Socket error", error)
> socks.GeneralProxyError: Socket error: 0x01: General SOCKS server failure
> ```
> 
> P.S. We could use Tor SOCKS5 proxy also for all downloads from GitHub.

Same for me...

Can you think of what could be the solution? open an issue to https://github.com/opentimestamps/opentimestamps-client maybe?

VajraOfIndra commented:
I updated this PR to v0.15.3. PTAL.

It's ready to be merged but they are some conflicts.. i tried to resolve them but I'm stuck. Could someone please resolve the conflicts in order for us to merge that PR? @kristapsk?

VajraOfIndra commented:
Closing because conflicting files.
Will reopen a new PR.
